bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/firmware/FirmwareUpdateServiceImpl.java

```diff
@@ -0,0 +1,556 @@
+
+    private final Map<ThingUID, FirmwareStatusInfo> firmwareStatusInfoMap = new ConcurrentHashMap<>();
+    private final Map<ThingUID, ProgressCallbackImpl> progressCallbackMap = new ConcurrentHashMap<>();
+
+    private final List<FirmwareUpdateHandler> firmwareUpdateHandlers = new CopyOnWriteArrayList<>();
+    private FirmwareRegistry firmwareRegistry;
+    private EventPublisher eventPublisher;
+    private TranslationProvider i18nProvider;
+    private LocaleProvider localeProvider;
+    private SafeCaller safeCaller;
+    private ConfigDescriptionValidator configDescriptionValidator;
+
+    private final Runnable firmwareStatusRunnable = new Runnable() {
+        @Override
+        public void run() {
+            logger.debug("Running firmware status check.");
+            for (FirmwareUpdateHandler firmwareUpdateHandler : firmwareUpdateHandlers) {
+                try {
+                    logger.debug("Executing firmware status check for thing with UID {}.",
+                            firmwareUpdateHandler.getThing().getUID());
+
+                    Firmware latestFirmware = getLatestSuitableFirmware(firmwareUpdateHandler.getThing());
+
+                    FirmwareStatusInfo newFirmwareStatusInfo = getFirmwareStatusInfo(firmwareUpdateHandler,
+                            latestFirmware);
+
+                    processFirmwareStatusInfo(firmwareUpdateHandler, newFirmwareStatusInfo, latestFirmware);
+                } catch (Exception e) {
+                    logger.debug("Exception occurred during firmware status check.", e);
+                }
+            }
+        }
+    };
+
+    @Activate
+    protected void activate(Map<String, Object> config) {
+        modified(config);
+    }
+
+    @Modified
+    protected synchronized void modified(Map<String, Object> config) {
+        logger.debug("Modifying the configuration of the firmware update service.");
+
+        if (!isValid(config)) {
+            return;
+        }
+
+        cancelFirmwareUpdateStatusInfoJob();
+
+        firmwareStatusInfoJobPeriod = config.containsKey(PERIOD_CONFIG_KEY) ? (Integer) config.get(PERIOD_CONFIG_KEY)
+                : firmwareStatusInfoJobPeriod;
+        firmwareStatusInfoJobDelay = config.containsKey(DELAY_CONFIG_KEY) ? (Integer) config.get(DELAY_CONFIG_KEY)
+                : firmwareStatusInfoJobDelay;
+        firmwareStatusInfoJobTimeUnit = config.containsKey(TIME_UNIT_CONFIG_KEY)
+                ? TimeUnit.valueOf((String) config.get(TIME_UNIT_CONFIG_KEY))
+                : firmwareStatusInfoJobTimeUnit;
+
+        if (!firmwareUpdateHandlers.isEmpty()) {
+            createFirmwareUpdateStatusInfoJob();
+        }
+    }
+
+    @Deactivate
+    protected void deactivate() {
+        cancelFirmwareUpdateStatusInfoJob();
+        firmwareStatusInfoMap.clear();
+        progressCallbackMap.clear();
+    }
+
+    /**
+     * Returns the {@link FirmwareStatusInfo} for the thing having the given thing UID.
```

Comment:
maybe we can remove the javadoc here

Reply:
Yes, I found all such places (where I copy pasted javadoc during the interface separation) and cleaned the odd javadoc.